pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,1029 @@
+use sp_std::{
+	cmp::{max, min},
+	vec,
+};
+#[cfg(test)]
+mod tests;
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+mod benchmarks;
+
+pub mod types;
+pub mod weights;
+
+use weights::WeightInfo;
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+
+use crate::types::*;
+
+type BlockNumberFor<T> = <T as frame_system::Config>::BlockNumber;
+
+pub const RETRY_TO_SEARCH_FOR_FREE_BLOCK: u32 = 5;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use codec::HasCompact;
+	use frame_support::traits::Contains;
+	use frame_support::weights::WeightToFee;
+
+	use frame_system::pallet_prelude::OriginFor;
+	use hydra_dx_math::ema::EmaPrice;
+	use hydradx_traits::pools::SpotPriceProvider;
+	use hydradx_traits::PriceOracle;
+	use orml_traits::NamedMultiReservableCurrency;
+	use sp_runtime::DispatchError::BadOrigin;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T>
+	where
+		<T as pallet_route_executor::Config>::AssetId: From<<T as pallet::Config>::Asset>,
+		<T as pallet_route_executor::Config>::Balance: From<u128>,
+		u128: From<<T as pallet_route_executor::Config>::Balance>,
+	{
+		fn on_initialize(current_blocknumber: T::BlockNumber) -> Weight {
+			let mut weight = <T as pallet::Config>::WeightInfo::on_initialize_with_empty_block();
+
+			let mut random_generator = T::RandomnessProvider::generator();
+
+			let mut schedule_ids: Vec<ScheduleId> = ScheduleIdsPerBlock::<T>::get(current_blocknumber).to_vec();
+
+			if !schedule_ids.is_empty() {
+				Self::deposit_event(Event::ExecutionsStarted {
+					block: current_blocknumber,
+				});
+			}
+
+			schedule_ids.sort_by_cached_key(|_| random_generator.gen::<u32>());
+			for schedule_id in schedule_ids {
+				let Some(schedule) = Schedules::<T>::get(schedule_id) else {
+					//We cant terminate here as there is no schedule information to do so
+					continue;
+				};
+
+				let next_execution_block =
+					match Self::prepare_schedule(current_blocknumber, &mut weight, schedule_id, &schedule) {
+						Ok(block) => block,
+						Err(err) => {
+							if err == Error::<T>::PriceChangeIsBiggerThanMaxAllowed.into() {
+								//The schedule is replanned instead of terminated
+							} else {
+								Self::terminate_schedule(schedule_id, &schedule, err);
+							}
+							continue;
+						}
+					};
+
+				let trade_result = Self::execute_schedule(schedule_id, &schedule);
+
+				match trade_result {
+					Ok(_) => {
+						if let Err(err) = Self::replan_or_complete(schedule_id, &schedule, next_execution_block) {
+							Self::terminate_schedule(schedule_id, &schedule, err);
+							continue;
+						}
+					}
+					Err(err) => {
+						Self::deposit_event(Event::TradeFailed {
+							id: schedule_id,
+							who: schedule.owner.clone(),
+							error: err,
+						});
+
+						if T::ContinueOnErrors::contains(&err) {
```

Comment:
maybe or surely? Because if not really needed, I would keep the current way of handling this.

If you insist on it, I can add it, but it will probably increase the complexity and comes with additional time.

Reply:
i am not sure if we have that list to just detect slippage limits error (which are only ones there right now) or we plan to use it for anything else. functionally it is the same

Reply:
refactored, so we don't have the list of errors anymore as we check against the limit in the pallet itself, then retry